### ISSUE TITLE
How do Heterogeneous Graphs link prediction

### ISSUE THREAD

**Issue by liyongkang123** [dmlc/dgl#3447]
## ❓ Questions and Help
Hi, I'm doing Heterogeneous Graphs link prediction now, 
I follow the user guide  https://docs.dgl.ai/en/latest/guide/training-link.html
But here, after the model is trained, how do you get the edges predicted by the model?
My code is as follows:

```
import dgl
import dgl.nn as dglnn
import dgl.function as fn
import torch.nn as nn
import torch.nn.functional as F
import torch
import numpy as np

n_users = 1000
n_items = 500
n_follows = 3000
n_clicks = 5000
n_dislikes = 500
n_hetero_features = 10
n_user_classes = 5
n_max_clicks = 10

follow_src = np.random.randint(0, n_users, n_follows)
follow_dst = np.random.randint(0, n_users, n_follows)
click_src = np.random.randint(0, n_users, n_clicks)
click_dst = np.random.randint(0, n_items, n_clicks)
dislike_src = np.random.randint(0, n_users, n_dislikes)
dislike_dst = np.random.randint(0, n_items, n_dislikes)

hetero_graph = dgl.heterograph({
    ('user', 'follow', 'user'): (follow_src, follow_dst),
    ('user', 'followed-by', 'user'): (follow_dst, follow_src),
    ('user', 'click', 'item'): (click_src, click_dst),
    ('item', 'clicked-by', 'user'): (click_dst, click_src),
    ('user', 'dislike', 'item'): (dislike_src, dislike_dst),
    ('item', 'disliked-by', 'user'): (dislike_dst, dislike_src)})

hetero_graph.nodes['user'].data['feature'] = torch.randn(n_users, n_hetero_features)
hetero_graph.nodes['item'].data['feature'] = torch.randn(n_items, n_hetero_features)
hetero_graph.nodes['user'].data['label'] = torch.randint(0, n_user_classes, (n_users,))
hetero_graph.edges['click'].data['label'] = torch.randint(1, n_max_clicks, (n_clicks,)).float()
hetero_graph.nodes['user'].data['train_mask'] = torch.zeros(n_users, dtype=torch.bool).bernoulli(0.6)
hetero_graph.edges['click'].data['train_mask'] = torch.zeros(n_clicks, dtype=torch.bool).bernoulli(0.6)

class HeteroDotProductPredictor(nn.Module):
    def forward(self, graph, h, etype):
        with graph.local_scope():
            graph.ndata['h'] = h
            graph.apply_edges(fn.u_dot_v('h', 'h', 'score'), etype=etype)
            return graph.edges[etype].data['score']

def construct_negative_graph(graph, k, etype):
    utype, _, vtype = etype
    src, dst = graph.edges(etype=etype)
    neg_src = src.repeat_interleave(k)
    neg_dst = torch.randint(0, graph.num_nodes(vtype), (len(src) * k,))
    return dgl.heterograph(
        {etype: (neg_src, neg_dst)},
        num_nodes_dict={ntype: graph.num_nodes(ntype) for ntype in graph.ntypes})


class RGCN(nn.Module):
    def __init__(self, in_feats, hid_feats, out_feats, rel_names):
        super().__init__()
        self.conv1 = dglnn.HeteroGraphConv({
            rel: dglnn.GraphConv(in_feats, hid_feats)
            for rel in rel_names}, aggregate='sum')
        self.conv2 = dglnn.HeteroGraphConv({
            rel: dglnn.GraphConv(hid_feats, out_feats)
            for rel in rel_names}, aggregate='sum')

    def forward(self, graph, inputs):
        h = self.conv1(graph, inputs)
        h = {k: F.relu(v) for k, v in h.items()}
        h = self.conv2(graph, h)
        return h

class Model(nn.Module):
    def __init__(self, in_features, hidden_features, out_features, rel_names):
        super().__init__()
        self.sage = RGCN(in_features, hidden_features, out_features, rel_names)
        self.pred = HeteroDotProductPredictor()
    def forward(self, g, neg_g, x, etype):
        h = self.sage(g, x)
        return self.pred(g, h, etype), self.pred(neg_g, h, etype)

def compute_loss(pos_score, neg_score):
    n_edges = pos_score.shape[0]
    return (1 - pos_score.unsqueeze(1) + neg_score.view(n_edges, -1)).clamp(min=0).mean()

k = 5 
model = Model(10, 20, 5, hetero_graph.etypes)
user_feats = hetero_graph.nodes['user'].data['feature']
item_feats = hetero_graph.nodes['item'].data['feature']
node_features = {'user': user_feats, 'item': item_feats}
opt = torch.optim.Adam(model.parameters())
for epoch in range(50):
    negative_graph = construct_negative_graph(hetero_graph, k, ('user', 'click', 'item'))
    pos_score, neg_score = model(hetero_graph, negative_graph, node_features, ('user', 'click', 'item'))
    loss = compute_loss(pos_score, neg_score)
    opt.zero_grad()
    loss.backward()
    opt.step()
    print(loss.item())
    print(pos_score.shape)
```

After training, the node representation can be obtained via:
`node_embeddings = model.sage(graph, node_features)`

I get the node representation, but I don't know how to continue using it.
Could you please provide a case of obtaining the connection relation between nodes?

Thank you very much for your help.

**Comment by BarclayII** [comment]
I have added a tutorial section on evaluating link prediction models on link prediction task in https://github.com/dmlc/dgl/pull/3463.  Does it address your problem?

Also @tanvirbKash for #3439.

**Comment by liyongkang123** [comment]
> I have added a tutorial section on evaluating link prediction models on link prediction task in #3463. Does it address your problem?
> 
> Also @tanvirbKash for #3439.



> I have added a tutorial section on evaluating link prediction models on link prediction task in #3463. Does it address your problem?
> 
> Also @tanvirbKash for #3439.

Thank you very much for your answer. I have completed the link inference task on my heterograph, using AUC to evaluate the performance of the model. But the guide you gave is about homogenous graphs, so it needs to be modified a bit. I did not use random batch sampling in your code.
Because in the evaluation section, test_graph is a homogenous graph and the original is a heterogeneous graph, this results in the following:

For step, (input_nodes, pair_graph, MFGS) in enumerate(TQ):

Error here, need to specify the type of edge. But I don't know how to change the code here

It would be nice if you had the time to tailor it specifically to the heterograph. Thank you very much for your help!

**Comment by liyongkang123** [comment]
> I have added a tutorial section on evaluating link prediction models on link prediction task in #3463. Does it address your problem?
> 
> Also @tanvirbKash for #3439.

Hello, I found a bug in your code, in Evaluating Performance with Link Prediction, line 399. What you wrote is 
`# Positive pairs
test_pos_src, test_pos_dst = graph.edges()
# Negative pairs
test_neg_src = test_pos_src
test_neg_dst = torch.randint(0, graph.num_nodes(), (graph.num_edges(),))


######################################################################
# First you need to construct a graph for ``dgl.dataloading.EdgeDataLoader``
# to iterate on, i.e. with the testing node pairs as edges.
# You also need to label the edges, 1 if positive and 0 if negative.
#

test_src = torch.cat([test_pos_src, test_neg_src])
test_dst = torch.cat([test_neg_src, test_neg_dst])`


but, it should be 
`test_dst = torch.cat([test_pos_dst , test_neg_dst])`

Thanks a lot!

**Comment by uergash1** [comment]
Hi @liyongkang123 
I am also struggling in heterogenous link prediction, and make case is very similar with the example you provided. 
Would you be able to share the solution you found?

**Comment by NewBoy6** [comment]
> Hi @liyongkang123 I am also struggling in heterogenous graph link prediction, and make case is very similar with the example you provided. Would you be able to share the solution you found?

Hello I'm sorry I can't make my code public at the moment, it would be better if you read this [guide ](https://docs.dgl.ai/en/latest/guide/training-link.html) and [tutorial](https://github.com/dmlc/dgl/pull/3463)

**Comment by liyongkang123** [comment]
> Hi @liyongkang123 I am also struggling in heterogenous graph link prediction, and make case is very similar with the example you provided. Would you be able to share the solution you found?

Hello I'm sorry I can't make my code public at the moment, it would be better if you read this [guide ](https://docs.dgl.ai/en/latest/guide/training-link.html) and [tutorial](https://github.com/dmlc/dgl/pull/3463)
